harvester/node-manager
# Add CloudInit controller

~~Depends on: #14~~

---

**Problem:**


A Harvester cluster operator can currently modify node configuration by SSH'ing into the node, and modifying the relevant Elemental cloud-init configuration file under `/oem`. However, this is incompatible with a GitOps approach to managing the state of the cluster, at least, without writing some code of their own.

**Solution:**

This PR adds a new CloudInit CRD, a controller, and an fsnotify watcher to harvester-node-manager. This allows cluster operators to layer a cloud-init configuration on top and to target specific nodes in the cluster with the `matchSelector` field in the cloud-init spec.

For example, to add SSH access to the Rancher user with my SSH key across every node in the cluster, I can represent this like so:

```
apiVersion: node.harvesterhci.io/v1beta1
kind: CloudInit
metadata:
  name: ssh-access
spec:
  matchSelector: {}
  filename: 99_ssh.yaml
  contents: |
    stages:
      network:
        - authorized_keys:
            rancher:
              - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPCUNsQEnKj0nl1GS07Qr5RDCCbCim4wu06hCzQZDmTk ckuehl@suselaptop
```

**Related Issue:** https://github.com/harvester/harvester/issues/3902

**Test plan:**

Much of the test plan appears in this PR as automated unit tests.

However, to confirm operation on a multi-node cluster, I ran these manual steps:

1. Apply the CRD (`kubectl apply -f ./manifests/crds/node.harvesterhci.io_cloudinits.yaml`)
2. Import a container built from this PR onto all nodes in the cluster (`ctr -n=k8s.io image import /home/rancher/harvester-node-manager.tar`)
3. Patch the Harvester ManagedChart (`kubectl patch managedchart harvester -n fleet-local -p='{"spec":{"paused":true}}' --type=merge`)
4. Edit the harvester-node-manager deployment to use the container imported in step 2 (`kubectl edit daemonset/harvester-node-manager -n harvester-system`)
5. Rollout the new container (`kubectl rollout restart daemonset/harvester-node-manager -n harvester-system`)
6. Wait for that to finish
7. Apply a CloudInit object (like the one listed above)
8. SSH in, confirm the file has been created under `/oem`.
